server/remote_asset/fetch_server/fetch_server.go

```diff
@@ -196,22 +198,37 @@ func mirrorToCache(ctx context.Context, bsClient bspb.ByteStreamClient, remoteIn
 	if rsp.StatusCode < 200 || rsp.StatusCode >= 400 {
 		return nil, status.UnavailableErrorf("failed to fetch %q: HTTP %s", uri, err)
 	}
+	// Download to a temp file to avoid running out of memory.
 	// TODO: Support cache uploads with unknown digest length, so that we can
-	// pipe directly from the HTTP response to the cache and avoid buffering
-	// the entire fetched contents in memory.
-	data, err := io.ReadAll(rsp.Body)
+	// pipe directly from the HTTP response to the cache.
+	f, err := scratchspace.CreateTemp("remote-asset-fetch-*")
 	if err != nil {
+		return nil, status.UnavailableErrorf("failed to create temp file for download: %s", err)
+	}
+	defer func() {
+		_ = f.Close()
+		if err := os.Remove(f.Name()); err != nil {
+			log.Errorf("Failed to remove temp file: %s", err)
+		}
+	}()
+	if _, err := io.Copy(f, rsp.Body); err != nil {
 		return nil, status.UnavailableErrorf("failed to read response body from %q: %s", uri, err)
 	}
-	blobDigest, err := digest.Compute(bytes.NewReader(data))
+	if _, err := f.Seek(0, 0); err != nil {
+		return nil, status.UnavailableErrorf("failed to seek temp file: %s", err)
+	}
+	blobDigest, err := digest.Compute(f)
 	if err != nil {
 		return nil, status.InternalErrorf("failed to compute digest: %s", err)
 	}
 	if expectedSHA256 != "" && blobDigest.Hash != expectedSHA256 {
 		return nil, status.InvalidArgumentErrorf("response body checksum for %q was %q but wanted %q", uri, blobDigest.Hash, expectedSHA256)
 	}
 	cacheRN := digest.NewCASResourceName(blobDigest, remoteInstanceName)
-	if _, err := cachetools.UploadFromReader(ctx, bsClient, cacheRN, bytes.NewReader(data)); err != nil {
+	if _, err := f.Seek(0, 0); err != nil {
+		return nil, status.UnavailableErrorf("failed to seek temp file: %s", err)
+	}
+	if _, err := cachetools.UploadFromReader(ctx, bsClient, cacheRN, f); err != nil {
```

Comment:
What you have is fine -- but if you wanted to close the file, upload it, then remove it, you could use cachetools.UploadFile instead.

Reply:
hm, I'm not a fan that UploadFile() tries to add things to the filecache, even though the app isn't configured with one today :/

Reply:
nvm, was looking at the wrong UploadFile(). Done -- looks much cleaner :)